### PR TITLE
Add logger implementation

### LINKED ISSUE #2: Implement logging interface
There are now a few examples for both host and guest implementations for an interface, so implementing the current logging interface _should_ be straightforward.

### PR DESCRIPTION
**Pull request by nitishm**
Closes #2 

Signed-off-by: Nitish Malhotra <nitishm@microsoft.com>

### CODE REVIEW

**Comment by nitishm** [comment]
@radu-matei I cannot seem to run the integration test for the logger. Not sure what I am missing?!

```terminal
dev (default) github/nitishm/wasi-experimental-toolkit on 🌱 nitishm/feat/2/implement-logger [📝🤷‍] is 📦 v0.1.0 via 🦀 v1.57.0 took 12s
🕙[ 16:04:00 ] [🔴 101] ✗  cargo test --tests wasi_log_tests
   Compiling wasi-experimental-toolkit v0.1.0 (/Users/nitishm/github/nitishm/wasi-experimental-toolkit)
    Finished test [unoptimized + debuginfo] target(s) in 7.97s
     Running unittests (target/debug/deps/wasi_experimental_toolkit-742a05f25e940d6e)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/integration.rs (target/debug/deps/integration-0f8c97e522fe45c3)

running 1 test
test wasi_log_tests::test_rust_log ... FAILED

failures:

---- wasi_log_tests::test_rust_log stdout ----
thread 'wasi_log_tests::test_rust_log' panicked at 'called `Result::unwrap()` on an `Err` value: unknown import: `wasi-log::info` has not been defined', tests/integration.rs:143:63
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    wasi_log_tests::test_rust_log

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 5 filtered out; finished in 0.31s

error: test failed, to rerun pass '--test integration'
```

**Comment by radu-matei** [comment]
Hi, @nitishm, thanks a lot for your contribution!

In the current model, there are two ways an interface can be satisfied:

- through another component — this requires us to _link_ the two together (and currently generate a statically linked Wasm module) before instantiating it
- through a host import — this requires us to setup the runtime to handle the imports from the guest.

And this is how the existing implementations are currently split, some are Wasm components, some are host implementations.

The tests simply import interfaces — so before instantiating them, we must either satisfy the imports through the host runtime, or link them with other Wasm components that implement those interfaces.

In your case, you implemented the log interface as a Wasm component, but did not link it before actually instantiating your tests — see the cache tests for how to set the linking for the test.

File: crates/wasi-log/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    version = "0.1.0"
+    edition = "2021"
+    authors = [ "Nitish Malhotra <nitishm@microsoft.com>" ]
+
+[lib]
+    crate-type = [ "cdylib" ]
+
+[dependencies]
+    anyhow           = "1.0"
+    wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", rev = "dab589f579b980c81b39c8348400b88a8c4d175a" }
```
Comment (radu-matei) [inline crates/wasi-log/Cargo.toml:12]: Could you please rebase on top of https://github.com/fermyon/wasi-experimental-toolkit/pull/5?

File: tests/modules/rust-logger/bindings.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
+mod wasi_log {
```
Comment (radu-matei) [inline tests/modules/rust-logger/bindings.rs:1]: This file doesn't need to be in source control.

Comment (nitishm) [reply tests/modules/rust-logger/bindings.rs]: removed

File: crates/wasi-log/Cargo.toml (hunk 1 above)
Comment (nitishm) [reply crates/wasi-log/Cargo.toml]: Will do

**Comment by nitishm** [comment]
> Hi, @nitishm, thanks a lot for your contribution!
> 
> In the current model, there are two ways an interface can be satisfied:
> 
> * through another component — this requires us to _link_ the two together (and currently generate a statically linked Wasm module) before instantiating it
> * through a host import — this requires us to setup the runtime to handle the imports from the guest.
> 
> And this is how the existing implementations are currently split, some are Wasm components, some are host implementations.
> 
> The tests simply import interfaces — so before instantiating them, we must either satisfy the imports through the host runtime, or link them with other Wasm components that implement those interfaces.
> 
> In your case, you implemented the log interface as a Wasm component, but did not link it before actually instantiating your tests — see the cache tests for how to set the linking for the test.

Thanks for explaining the subtle differences. I completely glazed over the fact that I was using rust-bindgen and not wasmtime bindgen in the log crate. Nevertheless it's been good exercise and learning opportunity to understand and try the two models. I opted for the wasmtime implementation. But if required I can either switch to the manual linkage option or add that to this existing option. LMK what you think, which option is better?

File: crates/log-wasmtime/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    authors = [ "Nitish Malhotra <nitishm@microsoft.com>" ]
+
+[lib]
+    doctest = false
+
+[dependencies]
+    anyhow               = "1.0"
+    env_logger = "0.9.0"
+    log = "0.4.0"
+    wit-bindgen-wasmtime = { git = "https://github.com/bytecodealliance/wit-bindgen", rev = "dab589f579b980c81b39c8348400b88a8c4d175a" }
```
Comment (radu-matei) [inline crates/log-wasmtime/Cargo.toml:14]: I think the revision used everywhere else in the repo is `2e654dc82b7f9331719ba617a36ed5967b2aecb0`

File: tests/modules/rust-log/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+version = "0.1.0"
+edition = "2021"
+authors = [ "Nitish Malhotra <nitishm@microsoft.com>" ]
+
+
+[lib]
+    crate-type = [ "cdylib" ]
+
+[dependencies]
+    wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", rev = "dab589f579b980c81b39c8348400b88a8c4d175a" }
```
Comment (radu-matei) [inline tests/modules/rust-log/Cargo.toml:12]: This should be revision `2e654dc82b7f9331719ba617a36ed5967b2aecb0`

File: crates/log-wasmtime/Cargo.toml (hunk 3 above)
Comment (radu-matei) [reply crates/log-wasmtime/Cargo.toml:14]: The current revision points to a version of WIT bindgen that uses Wasmtime v0.32, while the rest of the project has been updated to v0.33.

**Comment by radu-matei** [comment]
This mostl LGTM.
Manually updating the revision and running the tests passes:

```
running 1 test
[2022-01-12T17:13:46Z ERROR log_wasmtime] To err is human to rub it in is divine
test wasi_log_tests::test_rust_log ... ok
```

This is great, once you clean it up, update the revision, and (please) sign your commits, this should be good to go.

Thanks a lot!

File: tests/modules/rust-log/Cargo.toml (hunk 4 above)
Comment (nitishm) [reply tests/modules/rust-log/Cargo.toml:12]: Will update. I think I picked it up before the update was merged.

**Pull request merged**